misk-hibernate/src/main/kotlin/misk/hibernate/SchemaValidator.kt

```diff
@@ -0,0 +1,301 @@
+    }
+    return DatabaseDeclaration("databaseSchema", schemaTables)
+  }
+
+  private fun readDeclarationFromHibernate(metadata: Metadata): DatabaseDeclaration {
+    val hibernateTables = mutableListOf<TableDeclaration>()
+    val tableIt = metadata.collectTableMappings().iterator()
+    while (tableIt.hasNext()) {
+      val table = tableIt.next()
+      val tableName = table.name
+
+      val columnsIt = table.columnIterator
+      val columns = mutableListOf<ColumnDeclaration>()
+      while (columnsIt.hasNext()) {
+        val column = columnsIt.next() as Column
+        columns += ColumnDeclaration(column.name, column.isNullable)
+      }
+
+      hibernateTables += TableDeclaration(tableName, columns)
+    }
+    return DatabaseDeclaration("hibernateSchema", hibernateTables)
+  }
+
+  private fun reportString(): String {
+    val errorReport = StringBuilder("Failed Schema Validation: \n\n")
+    for (message in messages) {
+      if (message.error) errorReport.append("ERROR ") else errorReport.append("WARNING ")
+      errorReport.append("at ${message.path.joinToString(separator = ".")}:\n")
+      errorReport.append("  ${message.text.replace("\n", "\n  ")}\n")
+      errorReport.append("\n")
+    }
+    return errorReport.toString()
+  }
+
+  private fun validateDatabase(dbSchema: DatabaseDeclaration, hibernateSchema: DatabaseDeclaration) {
+    val (dbOnly, hibernateOnly, intersectionPairs) = splitChildren(
+        dbSchema.tables, hibernateSchema.tables)
+
+    validate(hibernateOnly.isEmpty()) {
+      "Database missing tables ${hibernateOnly.map { it.name }}"
+    }
+
+    checkWarning(dbOnly.isEmpty()) {
+      "Hibernate missing tables ${dbOnly.map { it.name }}"
+    }
+
+    for ((dbTable, hibernateTable) in intersectionPairs) {
+      withDeclaration(hibernateTable.snakeCaseName) {
+        validateTables(dbTable, hibernateTable)
+      }
+    }
+  }
+
+  private fun validateTables(dbTable: TableDeclaration, hibernateTable: TableDeclaration) {
+    val (dbOnly, hibernateOnly, intersectionPairs) = splitChildren(
+        dbTable.columns,
+        hibernateTable.columns)
+
+    validate(dbTable.snakeCaseName == dbTable.name) {
+      "Database table name \"${dbTable.name}\" should be in lower_snake_case"
+    }
+    validate(dbTable.name == hibernateTable.name) {
+      "Database table name \"${dbTable.name}\" should exactly match hibernate \"${hibernateTable.name}\""
+    }
+
+    validate(hibernateOnly.isEmpty()) {
+      "Database table \"${dbTable.name}\" is missing columns ${hibernateOnly.map { it.name }} found in hibernate \"${hibernateTable.name}\""
+    }
+
+    // TODO (maacosta) how strict should we be here? If `DEFAULT NULL` column exists in the Db and hibernate
+    //                does not know about it hibernate can still do writes to db. However if we are going to do
+    //                lookups on this column it might not work.. do we look at queries?.
+    validate(dbOnly.isEmpty()) {
```

Comment:
what do you mean with 
> However if we are going to do lookups on this column it might not work.. do we look at queries?.
How are you going to do lookups if the app doesn't know about it? 

This validation is dangerous if I change the schema of my table before deploying the app to production. I won't be able to restart the application until the new version of the app is deployed.